internal/controller/controllers/agent_controller.go

```diff
@@ -1598,6 +1598,33 @@ func (r *AgentReconciler) updateIfNeeded(ctx context.Context, log logrus.FieldLo
 		}
 	}
 
+	if agent.Spec.IgnitionEndpointHTTPHeaders != nil {
+		hostIgnitionEndpointHTTPHeaders := make(map[string]string)
+		if internalHost.IgnitionEndpointHTTPHeaders != "" {
+			if err = json.Unmarshal([]byte(internalHost.IgnitionEndpointHTTPHeaders), &hostIgnitionEndpointHTTPHeaders); err != nil {
+				log.WithError(err).Errorf("failed to unmarshal ignition endpoint HTTP headers for host %s infra-env %s", internalHost.ID.String(), internalHost.InfraEnvID.String())
+			}
+		}
+
+		if !reflect.DeepEqual(agent.Spec.IgnitionEndpointHTTPHeaders, hostIgnitionEndpointHTTPHeaders) {
+			funk.ForEach(agent.Spec.IgnitionEndpointHTTPHeaders, func(key, value string) {
+				params.HostUpdateParams.IgnitionEndpointHTTPHeaders = append(params.HostUpdateParams.IgnitionEndpointHTTPHeaders, &models.IgnitionEndpointHTTPHeadersParams{
+					Key:   swag.String(key),
+					Value: swag.String(value),
+				})
+			})
+			hostUpdate = true
+		}
+	} else {
+		if internalHost.IgnitionEndpointHTTPHeaders != "" {
+			hostUpdate = true
+			params.HostUpdateParams.IgnitionEndpointHTTPHeaders = append(params.HostUpdateParams.IgnitionEndpointHTTPHeaders, &models.IgnitionEndpointHTTPHeadersParams{
+				Key:   swag.String(""),
+				Value: swag.String(""),
```

Comment:
Sorry this is a bit confusing!

> Why empty key+value, instead of just unsetting?

Just unsetting it would look like there is no update because of [this check here](https://github.com/openshift/assisted-service/pull/5934/files#diff-747a429a0e12307ef65d754c83b6898ce3858b621005e1e7af6e420da1f1657cR6129):

```
func (b *bareMetalInventory) updateIgnitionEndpointHTTPHeaders(ctx context.Context, host *common.Host, ignitionEndpointHTTPHeadersList []*models.IgnitionEndpointHTTPHeadersParams, db *gorm.DB) error {
	log := logutil.FromContext(ctx, b.log)
	if ignitionEndpointHTTPHeadersList == nil {
		log.Infof("No request for ignition endpoint HTTP Headers update for host %s", host.ID)
		return nil
	}
```

So we set the empty key+value to actually unset it in the database.
Side note: This check is needed so that we don't update this every time host update is called.

> A couple lines above, you check if internalHost.IgnitionEndpointHTTPHeaders != "". Will this "if" condition work with an empty key+value?

Yes, `internalHost.IgnitionEndpointHTTPHeaders` is the DB column and if it's not empty then that means we need to unset it, otherwise if the agent doesn't specify the headers and it's empty then we don't need to do anything

Reply:
Let me know if you'd like me to clarify or reword!